File: backend/app/schemas/room.py
"""Physical room API contracts.

``rooms`` has no ``public_id`` and none is being added. ``room_number`` is the natural
identifier, backed by ``UNIQUE (hotel_id, room_number)``.

**That uniqueness is per HOTEL, not per room type.** The ``room-types`` segment in the URL is
therefore an assertion the room must satisfy, not part of its key: number ``101`` identifies
exactly one room in a hotel, whatever its type. A consequence worth stating plainly -- two
room types at one property cannot both own a room numbered ``101``, even though their URLs
read like separate collections.

``status`` is restricted to the five values the database CHECK constraint already allows. No
new state is invented, and none is booking-driven: ``status`` is the room's *current
operational* state, not its availability over time.
"""

from __future__ import annotations

import datetime as dt
import uuid
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

#: Mirrors RoomStatus in app.models.enums, which generates the database CHECK constraint.
#: Spelled out as a Literal so FastAPI documents the allowed values and rejects the rest at
#: the edge with a field-level message.
RoomStatusLiteral = Literal["available", "occupied", "cleaning", "maintenance", "out_of_order"]

#: "12A" and "P-3" are real room numbers, so this is TEXT, not an integer. Constrained to a
#: URL-safe shape because the value appears in the path.
RoomNumberField = Annotated[
    str, Field(min_length=1, max_length=20, pattern=r"^[A-Z0-9][A-Z0-9._-]*$")
]
FloorField = Annotated[int, Field(ge=-10, le=200)]


class RoomCreate(BaseModel):
    """Payload for POST .../rooms.

    Neither the hotel nor the room type appears in the body: both come from the URL path.
    Accepting them here as well would let the two disagree, and would hand the client a way
    to name a room type the path never authorised.
    """

    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

    room_number: RoomNumberField
    floor: FloorField | None = None
    status: RoomStatusLiteral = "available"
    notes: str | None = None
    is_active: bool = True

    @field_validator("room_number", mode="before")
    @classmethod
    def _upper(cls, value: object) -> object:
        """Canonicalise to upper case.

        PostgreSQL's unique constraint is case-sensitive, so without this ``101a`` and
        ``101A`` would be two different rooms in the same corridor.
        """
        return value.upper() if isinstance(value, str) else value


class RoomUpdate(BaseModel):
    """Payload for PATCH. Every field optional; omitted fields are left untouched.

    ``room_number`` is absent -- it is the URL identity, exactly as ``slug`` is for a hotel
    and ``code`` for a room type. The room type is absent too: moving a room between types
    would change its parent path, and the API deliberately never exposes ``room_type_id``.
    """

    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

    floor: FloorField | None = None
    status: RoomStatusLiteral | None = None
    notes: str | None = None
    is_active: bool | None = None


class RoomResponse(BaseModel):
    """What the API returns.

    Carries ``hotel_public_id`` and ``room_type_code`` so a client can rebuild the room's own
    URL from the payload alone. No internal key appears: not ``rooms.id``, not ``hotel_id``,
    not ``room_type_id``.
    """

    model_config = ConfigDict(from_attributes=True)

    hotel_public_id: uuid.UUID
    room_type_code: str
    room_number: str
    floor: int | None
    status: RoomStatusLiteral
    notes: str | None
    is_active: bool
    created_at: dt.datetime
    updated_at: dt.datetime


__all__ = ["RoomCreate", "RoomResponse", "RoomStatusLiteral", "RoomUpdate"]

File: frontend/src/types/room.ts
import type { DecimalString } from '@/types/analytics'

/**
 * The physical-room contracts, transcribed from `app.schemas.room` and confirmed against the
 * live OpenAPI document and real requests.
 *
 * Five facts that check established, each of which shaped the UI:
 *
 * 1. **A room has no `public_id`.** `room_number` is the natural identifier, and it appears
 *    in the URL. It is **upper-cased** by both the schema and the router, so `101a` and
 *    `101A` are one room -- verified: creating `s511c` produced `S511C`, and fetching
 *    `.../rooms/s511a` returned `S511A`.
 * 2. **Rooms are listed per room type, never per hotel.** The only list endpoint is
 *    `/hotels/{h}/room-types/{code}/rooms`. There is no hotel-wide room collection, which is
 *    why this screen asks which type to show rather than presenting one long list.
 * 3. **Room numbers are unique per HOTEL, not per type.** The schema is explicit and the
 *    server agrees: creating `S511A` under a *second* room type was a 409, not a success.
 *    The `room-types` segment is an assertion the room must satisfy, not part of its key.
 * 4. **`status` has five values and no transition rules.** Every ordering was accepted --
 *    `available` straight to `occupied`, `out_of_order` straight to `available` -- so the UI
 *    offers all five and invents no state machine. It is the room's *current operational*
 *    state, not its availability over time.
 * 5. **Neither the number nor the type can be changed by PATCH.** Both are 422s, verified:
 *    the number is the URL identity and moving a room between types would change its parent
 *    path.
 */

/**
 * Mirrors `RoomStatus` in `app.models.enums`, which generates the database CHECK.
 *
 * Sending anything else is a 422 naming all five, verified.
 */
export type RoomStatus = 'available' | 'occupied' | 'cleaning' | 'maintenance' | 'out_of_order'

/** One physical room, as the API returns it. */
export interface Room {
  readonly hotel_public_id: string
  /** The owning type's code. Carried so a client can rebuild the room's URL from the payload. */
  readonly room_type_code: string
  /** The identifier. Upper-case, 1..20 chars, `^[A-Z0-9][A-Z0-9._-]*$`. */
  readonly room_number: string
  /** `-10..200`, or null. Verified: -11 and 201 are both 422s. */
  readonly floor: number | null
  readonly status: RoomStatus
  readonly notes: string | null
  /**
   * Whether the room is in service.
   *
   * Distinct from `status`: a room can be `available` and inactive. The schema keeps them
   * apart and so does this application -- nothing here derives one from the other.
   */
  readonly is_active: boolean
  readonly created_at: string
  readonly updated_at: string
}

/**
 * `POST /hotels/{h}/room-types/{code}/rooms` -- the `RoomCreate` payload.
 *
 * Neither the hotel nor the room type appears: both come from the URL, and accepting them
 * here would let the two disagree. Sending `room_type_code` is a 422, verified.
 *
 * `status` defaults to `available` and `is_active` to `true`, in the schema and in the
 * columns alike.
 */
export interface RoomCreateRequest {
  readonly room_number: string
  readonly floor?: number | null
  readonly status?: RoomStatus
  readonly notes?: string | null
  readonly is_active?: boolean
}

/**
 * `PATCH .../rooms/{room_number}` -- the `RoomUpdate` payload.
 *
 * Four fields, and **not** `room_number` or `room_type_code`: both are 422s. An omitted field
 * is left untouched; an explicit null clears `floor` or `notes`. An empty payload is a 200
 * that changes nothing. All verified live.
 */
export interface RoomUpdateRequest {
  readonly floor?: number | null
  readonly status?: RoomStatus
  readonly notes?: string | null
  readonly is_active?: boolean
}

/**
 * A room type, as `GET /hotels/{h}/room-types` returns it.
 *
 * Read-only here. This stage manages **rooms**; the type collection is consumed because a
 * room cannot be listed without naming its type, and for the descriptive fields shown beside
 * the rooms. Creating or editing a type is not part of this screen.
 *
 * `base_price` and `size_sqm` are `Decimal`s and therefore arrive as **strings**, like every
 * other numeric of that kind in this API.
 */
export interface RoomType {
  readonly hotel_public_id: string
  readonly code: string
  readonly name: string
  readonly description: string | null
  readonly max_occupancy: number
  readonly standard_occupancy: number
  readonly bed_count: number
  readonly bed_configuration: string | null
  readonly size_sqm: DecimalString | null
  readonly base_price: DecimalString
  readonly currency: string
  readonly is_active: boolean
  readonly created_at: string
  readonly updated_at: string
}
